# PlantEuka ðŸŒ±ðŸŒ¿ðŸŒ»

**PlantEuka** is an extension of **Euka**, developed by Nicola et al. (https://doi.org/10.1111/2041-210X.14214), which uses a taxon-based pangenome graph for analyzing ancient environmental DNA (aeDNA). While **Euka** has shown great potential, it is limited to tetrapod and arthropod mitochondrial genomes (mitogenomes).

The **PlantEuka** project aims to develop a customized plant database module for **Euka**. This enhancement will improve the accuracy and representation of plant species, enabling the identification of plant species using a taxon-based pangenome graph.

The primary goal of this project is to enhance our understanding of plant biodiversity and their interactions in various environments. These improvements will allow **Euka** to provide more reliable data for ecological and biological research.

## How to Use the PlantEuka

Download the repository and follow the steps. Ensure all scripts are executed from the main directory of PlantEuka.

### download_genomes.py

Connect with NCBIâ€™s GenBank database, search and retrieve sequences matching the specified query string.
```bash
scripts/download_genomes.py <organelle>
```

### Taxonomy files 

Download taxonomy files, taxdump.tar.gz and nucl_gb.accession2taxid.gz from NCBI's FTP repository. 
```bash
wget -bqc -P other https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz
```
```bash
wget -bqc -P other https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/nucl_gb.accession2taxid.gz
```

### accession_taxid.sh

Match downloaded accession numbers with their corresponding taxon IDs (taxids) using files from the NCBI database.
```bash
scripts/accession_taxid.sh
```


### taxonomy_lineage.py

Extract taxonomy and lineage information from `nodes.dmp` and `names.dmp` files, creating `taxonomy.tsv` and `lineage.tsv`.
```bash
scripts/taxonomy_lineage.py
```

### sort_genomes.py

Organize genomic data into directories by taxonomic rank (genus, family, order), creating lists of sorted and unsorted accession numbers.
```bash
scripts/sort_genomes.py <organelle>
```

### clean_genomes.py

Scan genomic sequences for non-standard characters and replaces them with a standard â€™Nâ€™ for unknown nucleotides.
```bash
scripts/clean_genomes.py <organelle>
```

### Use EMBOSS Stretcher for pairwise alignment. The aim is to filter unwanted pairs and duplicates.

Create a list with all the possible pairs within a group.
```bash
pairwise_list_parallel.sh
```

Run each pair from the list with stretcher. 
```bash
tools/stretcher -asequence <seq1>.fasta -bsequence <seq2>.fasta -gapopen 16 -gapextend 4 -outfile <seq1_seq2_pair>.stretcher 
```

### Gather and filter results

Gather the results from the pairwise alignment into a single .tsv file.
```bash
scripts/pairwise_results.py  
```

Filter the sequences using specific criteria.
```bash
scripts/pairwise_filter.py 
```

**Criteria**

Similarity > 99% -> One of the pairs is discarded (removes duplicates) 

50% > Similarity > 90%

Gaps < 15%

Score > MaxScore/2 (MaxScore represents the highest score found per group)

### MAFFT for multiple sequence alignment

Merge, per group, the individual sequences for the MSA with MAFFT.
```bash
scripts/merge_genomes.py 
```

Run MAFFT. 
```bash
tools/mafft --auto --thread -1 <merged_sequences>.fasta > <alignment>.fasta
```

### Construct a Pangenome Graph 


Note: Still under development. More details will be added soon.


